buf/registry/module/v1beta1/branch_service.proto

```diff
@@ -98,3 +115,132 @@ message ListBranchesResponse {
+  // two BranchRefs refer to the same Branch.
+  repeated Value values = 1 [(buf.validate.field).repeated.min_items = 1];
+}
+
+message PushResponse {
+  // The Branches that were pushed to, in the order they appeared in the request.
+  repeated Branch branches = 1 [(buf.validate.field).repeated.min_items = 1];
+}
+
+message RebaseRequest {
+  // An operation to apply as part of the rebase.
+  message Operation {
+    // An operation to create a Branch.
+    message CreateBranch {
+      // The Module to create the Branch on.
+      ModuleRef module_ref = 1 [(buf.validate.field).required = true];
+      // The name of the Branch to create.
+      //
+      // If a Branch with this name already exists on the Module, this will result in an error.
+      string branch_name = 2 [
+        (buf.validate.field).required = true,
+        (buf.validate.field).string.max_len = 250
+      ];
+      // The ID of the Commit from which to start history. The history of the branch will
+      // reflect the history as if you called ListCommitHistory for this Commit ID.
+      //
+      // If this is empty, the created Branch will only have the Commits specified in
+      // add_commit_ids.
+      //
+      // This is primarily designed for use in forking a Branch instead of
+      // providing the new Branch's entire history.
+      string start_commit_id = 3 [(buf.validate.field).string.uuid = true];
+      // The IDs of the Commits to add to the Branch, in the order they should be added.
+      //
+      // If start_commit_id is specified, these will be appended to the Commit history
+      // that aleady exists up to start_commit_id.
+      //
+      // If start_commit_id is not specified, the newly-created Branch will have exactly
+      // these Commits, in this order.
+      repeated string add_commit_ids = 4 [(buf.validate.field).repeated.items = {
+        string: {uuid: true}
+      }];
+    }
+
+    // An operation to update an existing branch.
+    message UpdateBranch {
+      // The branch to update.
+      BranchRef branch_ref = 1 [(buf.validate.field).required = true];
+      // The ID of the Commit that the caller expects to be the head Commit on the Branch.
+      //
+      // If this ID does not match the ID of the head Commit on the Branch, an error is
+      // returned.
+      string expected_head_commit_id = 2 [
+        (buf.validate.field).required = true,
+        (buf.validate.field).string.uuid = true
+      ];
+      // The IDs of the Commits to remove from the Branch.
+      //
+      // Commits are removed before any Commits are added from add_commit_ids.
+      //
+      // Note that it is valid to have one of the remove_commit_ids be equal to the
+      // expected_head_commit_id. If this is the case, the Branch will have a different
+      // head Commit that add_commit_ids will be applied to.
+      repeated string remove_commit_ids = 3 [(buf.validate.field).repeated.items = {
```

Comment:
> Note that it is valid to have one of the remove_commit_ids be equal to the expected_head_commit_id. If this is the case, the Branch will have a different head Commit that add_commit_ids will be applied to.

This comment implies that we support removing commits _at the middle_ of the history, which is not true with the spec we're proposing. When removing, it's always commits at the tail of the branch, so it's not "_valid_ that one of the `remove_commit_ids` is the `expected_head_commit_id`", if we're removing commits, "the head is _required_ to be present as the last item in this array".

Reply:
You are right the current doc is a bit underspecified. In addition to documenting the point you made, I think we should also require/document that the order of commits matters (it is the order that they will be popped off the branch and at each step if the head commit id doesn't match the id trying to be popped off, the op fails).

Reply:
@unmultimedio is correct, the ID of the head Commit _must_ be present in `remove_commit_ids`.

To @nicksnyder's point, this doc used to say

> These commits must appear in order at the end of the Branch's history.

I think we should keep defining the expected order of `commits_ids_to_remove`. I would also expect this to match the order in which they appear in Branch history, not the order in which they will be removed. The order in which they will be removed is not defined by the user, it's defined by the system. We simply care that the state we're asserting is valid, the actual order of how history is updated is based on some business rules that may change over time.

In other words, we receive a request that says:

> my history looks like `(..., A, B, C]`, I would like it to be `(..., A, D]`

We only need to validate that request matches our current state, and follow our business rules to carry out that request. We can _define_ how this will happen, but I wouldn't want the user to drive this (and for us to validate the user's expectations of how this will happen).